merlion/models/forecast/deep_base.py

```diff
@@ -0,0 +1,300 @@
+                batch_size=config.batch_size,
+                target_seq_index=None,  # set None, we use config.target_seq_index later in the training
+                ts_encoding=config.ts_encoding,
+                start_token_len=config.start_token_len,
+                flatten=False,
+                shuffle=False,
+            )
+        else:
+            eval_dataset = evaluate_data
+
+        all_preds = []
+        all_trues = []
+        total_loss = []
+
+        for i, batch in enumerate(eval_dataset):
+            with torch.no_grad():
+                loss, outputs, y_true = self._get_batch_model_loss_and_outputs(batch)
+                pred = outputs.detach().cpu().numpy()
+                true = y_true.detach().cpu().numpy()
+
+                all_preds.append(pred)
+                all_trues.append(true)
+                total_loss.append(loss.item())
+
+        preds = np.concatenate(all_preds, axis=0)
+        trues = np.concatenate(all_trues, axis=0)
+
+        return preds, np.average(total_loss)
+
+    @property
+    def support_multivariate_output(self) -> bool:
+        """
+        Deep models support multivariate output by default.
+        """
+        return True
+
+    def _train(self, train_data: pd.DataFrame, train_config=None) -> pd.DataFrame:
+        config = self.config
+
+        # creating model before the training
+        self._create_model()
+        if config.use_gpu:
+            self.to_gpu()
+        else:
+            self.to_cpu()
+
+        total_data = copy.deepcopy(train_data)
+        data_size = len(total_data)
+
+        # splitting the data into training and valdation set
+        if config.validation_rate > 0:
+            training_data = total_data[: int(data_size * (1 - config.validation_rate))]
+            validation_data = total_data[int(data_size * (1 - config.validation_rate)) :]
+        else:
+            # if validation rate is zero, we use all training data as validation
+            training_data = copy.deepcopy(total_data)
+            validation_data = copy.deepcopy(total_data)
+
+        training_dataset = RollingWindowDataset(
+            training_data,
+            n_past=config.n_past,
+            n_future=config.max_forecast_steps,
+            batch_size=config.batch_size,
+            target_seq_index=None,  # have to set None, we use config.target_seq_index later in the training, if not this is a bug
+            ts_encoding=config.ts_encoding,
+            start_token_len=config.start_token_len,
+            flatten=False,
+            shuffle=True,
+        )
+
+        validation_dataset = RollingWindowDataset(
+            validation_data,
+            n_past=config.n_past,
+            n_future=config.max_forecast_steps,
+            batch_size=config.batch_size,
+            target_seq_index=None,
+            ts_encoding=config.ts_encoding,
+            start_token_len=config.start_token_len,
+            flatten=False,
+            shuffle=False,
+        )
+
```

Comment:
Can you modify `RollingWindowDataset` to accept parameters `valid_frac` (float) and `validation` (bool)? I envision the following behavior:

- set `n_valid = math.ceil(len(self) * self.valid_frac)`
- use the last `n_valid` windows for validation & the remaining windows for training
- if `validation` is `False`, we only provide access to the training windows. If `validation` is `True`, we only provide access to the validation windows. make sure that `n_points` and `__getitem__` are updated accordingly. 
- if `valid_frac` is 0  or 1, then we iterate over the entire dataset, regardless of whether `validation` is `True` or `False`

In this way, we can use all of the available data more effectively.